src/libraries/Microsoft.Extensions.Logging/gen/LoggerMessageGenerator.Parser.cs

```diff
@@ -0,0 +1,661 @@
+                }
+
+                INamedTypeSymbol logLevelSymbol = _compilation.GetTypeByMetadataName("Microsoft.Extensions.Logging.LogLevel");
+                if (logLevelSymbol == null)
+                {
+                    // nothing to do if this type isn't available
+                    return Array.Empty<LoggerClass>();
+                }
+
+                INamedTypeSymbol exceptionSymbol = _compilation.GetTypeByMetadataName("System.Exception");
+                if (exceptionSymbol == null)
+                {
+                    Diag(DiagnosticDescriptors.MissingRequiredType, null, "System.Exception");
+                    return Array.Empty<LoggerClass>();
+                }
+
+                INamedTypeSymbol enumerableSymbol = _compilation.GetTypeByMetadataName("System.Collections.IEnumerable");
+                if (enumerableSymbol == null)
+                {
+                    Diag(DiagnosticDescriptors.MissingRequiredType, null, "System.Collections.IEnumerable");
+                    return Array.Empty<LoggerClass>();
+                }
+
+                INamedTypeSymbol stringSymbol = _compilation.GetTypeByMetadataName("System.String");
+                if (stringSymbol == null)
+                {
+                    Diag(DiagnosticDescriptors.MissingRequiredType, null, "System.String");
+                    return Array.Empty<LoggerClass>();
+                }
+
+                var results = new List<LoggerClass>();
+                var ids = new HashSet<int>();
+
+                // we enumerate by syntax tree, to minimize the need to instantiate semantic models (since they're expensive)
+                foreach (var group in classes.GroupBy(x => x.SyntaxTree))
+                {
+                    SemanticModel? sm = null;
+                    foreach (ClassDeclarationSyntax classDec in group)
+                    {
+                        // stop if we're asked to
+                        _cancellationToken.ThrowIfCancellationRequested();
+
+                        LoggerClass? lc = null;
+                        string nspace = string.Empty;
+                        string? loggerField = null;
+                        bool multipleLoggerFields = false;
+
+                        ids.Clear();
+                        foreach (var member in classDec.Members)
+                        {
+                            var method = member as MethodDeclarationSyntax;
+                            if (method == null)
+                            {
+                                // we only care about methods
+                                continue;
+                            }
+
+                            foreach (AttributeListSyntax mal in method.AttributeLists)
+                            {
+                                foreach (AttributeSyntax ma in mal.Attributes)
+                                {
+                                    sm ??= _compilation.GetSemanticModel(classDec.SyntaxTree);
```

Comment:
Do we need to check if `sm` is null on every attribute? This is going to be executed on every attribute found in the attribute lists.

Reply:
cc @geeknoid

Reply:
Feel free to address on a follow up PR.

Reply:
I think it makes sense to move this one layer up. As it is, it only gets the semantic model if it finds a member with an attribute on it. Since the vast majority of types end up having no attributes on them, it's a good economy not to materialize the semantic model for all those cases.